core/src/main/java/org/neo4j/ogm/metadata/ClassInfo.java

```diff
@@ -991,21 +990,18 @@ public void registerIdGenerationStrategy(IdStrategy strategy) {
         }
     }
 
-    public MethodInfo postLoadMethodOrNull() {
+    public synchronized MethodInfo postLoadMethodOrNull() {
         if (isPostLoadMethodMapped) {
             return postLoadMethod;
         }
-        if (!isPostLoadMethodMapped) {
-            for (MethodInfo methodInfo : methodsInfo().methods()) {
-                if (methodInfo.hasAnnotation(PostLoad.class.getName())) {
-                    isPostLoadMethodMapped = true;
-                    postLoadMethod = methodInfo;
-                    return postLoadMethod;
-                }
-            }
-            isPostLoadMethodMapped = true;
+
+        Collection<MethodInfo> possiblePostLoadMethods = methodsInfo.findMethodInfoBy(methodInfo -> methodInfo.hasAnnotation(PostLoad.class));
+        if(possiblePostLoadMethods.size() > 1) {
+            throw new MetadataException(String.format("Cannot have more than one post load method annotated with @PostLoad for class '%s'", this.className));
         }
-        return null;
+
+        this.postLoadMethod = possiblePostLoadMethods.stream().findFirst().orElse(null);
```

Comment:
Setting an instance variable and returning it afterwards looks like either the calling part uses this method wrong and/or this method has multiple purposes.
Additionally the mixed usage of `this` and without it.

Reply:
I kept the method the way it was (mainly), but indeed, better split it like the other `initXXX` methods.